### PR TITLE
Add second derivatives

### PR DESCRIPTION
**Pull request by frostedoyster**

### CODE REVIEW

**Comment by frostedoyster** [comment]
@Luthaf ready for your review!

File: benchmarks/cpp/benchmark.cpp [hunk 1]
```diff
@@ -18,11 +18,13 @@
 using namespace sphericart;
 // shorthand for all-past-1 generic sph only
 template<typename DTYPE>
-inline void compute_generic(int n_samples, int l_max, DTYPE *prefactors, DTYPE *xyz, DTYPE *sph, DTYPE *dsph, DTYPE* buffers) {
+inline void compute_generic(int n_samples, int l_max, DTYPE *prefactors, DTYPE *xyz, DTYPE *sph, DTYPE *dsph, DTYPE* ddsph, DTYPE* buffers) {
```
[the remark this answers is not in the document]
Comment (Luthaf) [reply benchmarks/cpp/benchmark.cpp:21]: Alternatively, we could name these `sph`, `dsph_dxyz` and `d2sph_d2xyz`

[the remark this answers is not in the document]
Comment (frostedoyster) [reply benchmarks/cpp/benchmark.cpp:21]: I agree. `sph`, `gradients`, `hessians` would also be a possibility. Let's leave it for an improved documentation PR maybe (which I think we need anyway)

File: benchmarks/cpp/benchmark.cpp [hunk 2]
```diff
@@ -18,11 +18,13 @@
...
     if (dsph==nullptr) {
-        generic_sph<DTYPE, false, false, 1>(xyz, sph, dsph, n_samples, l_max, prefactors, buffers);
+        generic_sph<DTYPE, false, false, false, 1>(xyz, sph, dsph, ddsph, n_samples, l_max, prefactors, buffers);
```
[the remark this answers is not in the document]
Comment (frostedoyster) [reply benchmarks/cpp/benchmark.cpp:23]: I wouldn't bother because this is not part of the interface anymore... It's only in the benchmark file. Our interface now has three different functions:
`compute`
`compute_with_gradients`
`compute_with_hessians`
and they only take the arguments that are relevant, if I'm not mistaken

File: python/src/sphericart/spherical_harmonics.py [hunk 3]
```diff
@@ -206,3 +206,102 @@ def compute_with_gradients(self, xyz: np.ndarray) -> Tuple[np.ndarray, np.ndarra
             )
 
         return sph, dsph
+
+    def compute_with_hessians(
```
[the remark this answers is not in the document]
Comment (frostedoyster) [reply python/src/sphericart/spherical_harmonics.py:210]: Michele and I talked about this, and, although it sounds wrong, it is correct mathematically because each Ylm is a function with its own gradient and Hessian, so it makes sense to use `gradients` and `Hessians`.

File: python/src/sphericart/spherical_harmonics.py [hunk 4]
```diff
@@ -206,3 +206,102 @@ def compute_with_gradients(self, xyz: np.ndarray) -> Tuple[np.ndarray, np.ndarra
...
+        coordinates are in the ``xyz`` array, together with their Cartesian
+        derivatives and second derivatives.
+
+        >>> import numpy as np
+        >>> import sphericart as sc
+        >>> sh = sc.SphericalHarmonics(l_max=8, normalized=False)
+        >>> xyz = np.random.normal(size=(10,3))
+        >>> sh_values, sh_grads, sh_hessians = sh.compute_with_hessians(xyz)
+        >>> sh_hessians.shape
+        (10, 3, 3, 81)
```
[the remark this answers is not in the document]
Comment (frostedoyster) [reply python/src/sphericart/spherical_harmonics.py:224]: Very good idea

File: sphericart-torch/src/autograd.cpp [hunk 5]
```diff
@@ -262,24 +316,79 @@ torch::autograd::variable_list SphericalHarmonicsAutograd::backward(
...
+
+torch::autograd::variable_list SphericalHarmonicsAutogradBackward::backward(
+    torch::autograd::AutogradContext *ctx,
+    torch::autograd::variable_list grad_2_outputs
+) {
+    auto saved_variables = ctx->get_saved_variables();
+    if (saved_variables.size() == 0) {
+        // No saved variables, meaning that the user called the double backward without specifying 
+        // that they want to do backward second derivatives when creating the class.
+        throw std::runtime_error("internal error: please specify backward_second_derivatives=True at class instantiation if you want to perform double backpropagation");
```
[the remark this answers is not in the document]
Comment (frostedoyster) [reply sphericart-torch/src/autograd.cpp:363]: True, well spotted

File: sphericart/include/templates.hpp [hunk 6]
```diff
@@ -83,93 +91,159 @@ void compute_sph_prefactors(int l_max, T *factors) {
...
     [[maybe_unused]] auto y2 = y * y;
     [[maybe_unused]] auto z2 = z * z;
-    [[maybe_unused]] T ir=0.0; // 1/r, it is only computed and used if we normalize the input vector
+    [[maybe_unused]] T ir = 0.0; // 1/r, it is only computed and used if we normalize the input vector
 
-    if constexpr(NORMALIZED) {
-        ir = 1/std::sqrt(x2+y2+z2);
-        x*=ir; y*=ir; z*=ir;
-        x2 = x*x; y2=y*y; z2=z*z;
+    if constexpr (NORMALIZED)
```
[the remark this answers is not in the document]
Comment (frostedoyster) [reply sphericart/include/templates.hpp:160]: I think Nick changed it.  @nickjbrowning what did you use to format the file and do you think we should enforce the format? I don't have any strong feelings about it

File: sphericart/include/templates.hpp [hunk 7]
```diff
@@ -83,93 +91,159 @@ void compute_sph_prefactors(int l_max, T *factors) {
...
         T *dxsph_i = dsph_i;
         T *dysph_i = dxsph_i + size_y;
         T *dzsph_i = dysph_i + size_y;
         HARDCODED_SPH_DERIVATIVE_MACRO(HARDCODED_LMAX, x, y, z, x2, y2, z2, sph_i, dxsph_i, dysph_i, dzsph_i, DUMMY_SPH_IDX);
 
-        if constexpr(NORMALIZED) {
+        if constexpr (DO_SECOND_DERIVATIVES)
+        {
+            // set each second derivative pointer to the appropriate place
+            T *dxdxsph_i = ddsph_i;
```
[the remark this answers is not in the document]
Comment (frostedoyster) [reply sphericart/include/templates.hpp]: I agree. I've applied the first one throughout

File: python/src/sphericart/spherical_harmonics.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Luthaf) [reply python/src/sphericart/spherical_harmonics.py:210]: ok, fair enough

File: sphericart/include/templates.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (Luthaf) [reply sphericart/include/templates.hpp:160]: See #57 for the follow up to this discussion

File: sphericart-torch/python/tests/test_nn.py [hunk 8]
```diff
@@ -31,9 +31,11 @@ def forward(self, positions):
                 retain_graph=True,
                 create_graph=True,
             )[0]
+            positions.requires_grad = False  # If not, further gradients with respect to xyz will be requested in backward()
```
Comment (Luthaf) [inline sphericart-torch/python/tests/test_nn.py]: this is weird. Changing require_grad at this point should not change the graph

Comment (frostedoyster) [reply sphericart-torch/python/tests/test_nn.py]: Thanks for confirming. I've removed it

**Review by Luthaf — APPROVED** [review]
